lib/Dialect/SV/Transforms/HWDeclareTypedecls.cpp

```diff
@@ -0,0 +1,87 @@
+//===- HWDeclareTypedecls.cpp - HW DeclareTypedecls Pass ------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "SVPassDetail.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/HW/HWTypes.h"
+#include "circt/Dialect/SV/SVPasses.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/SymbolTable.h"
+
+using namespace circt;
+using namespace hw;
+
+namespace {
+struct HWDeclareTypedeclsPass
+    : public sv::HWDeclareTypedeclsBase<HWDeclareTypedeclsPass> {
+  void runOnOperation() override;
+
+private:
+  void visitOperation(Operation *op);
+  void visitType(Operation *op, Type type);
+  ModuleOp parentModule;
+  OpBuilder *builder;
+};
+} // namespace
+
+void HWDeclareTypedeclsPass::visitOperation(Operation *op) {
+  for (auto type : op->getOperandTypes())
+    visitType(op, type);
+  for (auto type : op->getResultTypes())
+    visitType(op, type);
+  for (auto &region : op->getRegions())
+    for (auto &block : region.getBlocks())
+      for (auto type : block.getArgumentTypes())
+        visitType(op, type);
+}
+
+void HWDeclareTypedeclsPass::visitType(Operation *op, Type type) {
+  // Look for any TypeAliasTypes.
+  auto alias = type.dyn_cast<TypeAliasType>();
+  if (!alias)
+    return;
+
+  // If no scope exists, create one in the parent module.
+  StringRef scopeName = alias.getRef().getRootReference();
+  TypeScopeOp scope = parentModule.lookupSymbol<TypeScopeOp>(scopeName);
+  if (!scope) {
+    builder->setInsertionPointToStart(parentModule.getBody());
+    scope = builder->create<TypeScopeOp>(op->getLoc(), scopeName);
+    scope.body().emplaceBlock();
+  }
+
+  // If no typedecl exists, create one in the scope.
+  StringRef symbolName = alias.getRef().getLeafReference();
+  TypedeclOp typeDecl = scope.lookupSymbol<TypedeclOp>(symbolName);
+  if (!typeDecl) {
+    // TODO: The insertion point should generate type aliases of type aliases in
```

Comment:
This is where the logic from https://github.com/llvm/circt/pull/1077#discussion_r634825610 can be implemented. For now, I've left it as a todo, as this can handle the simple cases of giving a pretty name to arrays and structs. I can add that support here if it's deemed necessary.

Reply:
I think that's fine for a first pass. Once this gets merged, please create an issue for this improvement.